notebooks/04_model_select_and_optimize/hypertune_aml_wide_and_deep_quickstart.ipynb

```diff
@@ -0,0 +1,1109 @@
+      "Uploading aml_data/movielens_100k_valid.pkl\n",
+      "Uploaded aml_data/movielens_100k_valid.pkl, 1 files out of an estimated total of 2\n",
+      "Uploaded aml_data/movielens_100k_train.pkl, 2 files out of an estimated total of 2\n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "$AZUREML_DATAREFERENCE_2a7a3cb4cf38480abd3eda88df08b2e1"
+      ]
+     },
+     "execution_count": 32,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "DATA_DIR = 'aml_data'\n",
+    "\n",
+    "os.makedirs(DATA_DIR, exist_ok=True)\n",
+    "\n",
+    "TRAIN_FILE_NAME = \"movielens_\" + MOVIELENS_DATA_SIZE + \"_train.pkl\"\n",
+    "train.to_pickle(os.path.join(DATA_DIR, TRAIN_FILE_NAME))\n",
+    "VALID_FILE_NAME = \"movielens_\" + MOVIELENS_DATA_SIZE + \"_valid.pkl\"\n",
+    "valid.to_pickle(os.path.join(DATA_DIR, VALID_FILE_NAME))\n",
+    "\n",
+    "# Note, all the files under DATA_DIR will be uploaded to the data store\n",
+    "ds = ws.get_default_datastore()\n",
+    "ds.upload(\n",
+    "    src_dir=DATA_DIR,\n",
+    "    target_path='data',\n",
+    "    overwrite=True,\n",
+    "    show_progress=True\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 4. Prepare Training Scripts\n",
+    "Next step is to prepare scripts that AzureML Hyperdrive will use to train and evaluate models with selected hyperparameters. We re-use our [Wide-Deep Quickstart notebook](../00_quick_start/wide_deep_movielens.ipynb) for that. To run the model notebook from the Hyperdrive Run, all we need is to prepare an [entry script](../../reco_utils/aml/wide_deep.py) which parses the hyperparameter arguments, passes them to the notebook, and records the results of the notebook to AzureML Run logs by using `papermill`. Hyperdrive uses the logs to track the performance of each hyperparameter-set and finds the best performed one.  \n",
+    "\n",
+    "Here is a code snippet from the [entry script](../../reco_utils/aml/wide_deep.py):\n",
+    "```\n",
+    "import argparse\n",
+    "import papermill as pm\n",
+    "from azureml.core import Run\n",
+    "run = Run.get_context()\n",
+    "...\n",
+    "parser = argparse.ArgumentParser()\n",
+    "...\n",
+    "parser.add_argument('--dnn-optimizer', type=str, dest='dnn_optimizer', ...\n",
+    "parser.add_argument('--dnn-optimizer-lr', type=float, dest='dnn_optimizer_lr', ...\n",
+    "...\n",
+    "pm.execute_notebook(\n",
+    "    \"../../notebooks/00_quick_start/wide_deep_movielens.ipynb\",\n",
+    "    OUTPUT_NOTEBOOK,\n",
+    "    parameters=params,\n",
+    "    kernel_name='python3',\n",
+    ")\n",
+    "...\n",
+    "```"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Prepare all the necessary scripts which will be loaded to our Hyperdrive Experiment Run\n",
+    "SCRIPT_DIR = 'aml_script'\n",
```

Comment:
Could we avoid copying the entire reco_utils subtree under a new directory?

Reply:
Btw I tried `os.symlink()` as a potential workaround but I get a "no privilege" error on Windows.

Reply:
You can simply pass the repo's base path "Recommender/" instead of copying into a temp path and use the temp path:
* pros: our aml_notebook's script loading cell will become super simple
* cons: all the heavy notebooks and not-necessary tests, non-python files will be uploaded

Otherwise, you have to copy each files the script has dependency on and their folders.

Which why do we prefer?

Reply:
Neither of the two :-) What I am suggesting is to avoid copying a subtree.  
The way you do it is fine but instead of `shutil.copytree` you copy a single file i.e. wide_deep.py.

Reply:
wide_deep.py calls the model notebook, and the model notebook have dependency on our repo... e.g. tf_utils, movielens, pandas_df_utils, python_splitters, python_evaluation, constants. So better to just upload all repo (except tests and notebooks) to AzureML workspace blob.

Reply:
Why? This is not necessary. It works without doing this.